articles/102_dependency_groups.md

```diff
@@ -0,0 +1,164 @@
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+published: true
+categories: Overview
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Preface
+
+In the ROS ecosystem a ROS package provides information about its dependencies in the manifest.
+The latest format is specified in the [REP 140](http://www.ros.org/reps/rep-0140.html).
+It supports different kind of dependencies (`build`, `exec`, `test`, `doc`, etc.).
+But each dependency is mandatory and needs to satisfied on its own.
+
+This article describes the concept of "dependency groups" which could be used for different purposes.
+In terms of terminology a dependency group is a set of dependencies identified by the name of the group.
+For such a group a different semantic can then be applied (instead of just satisfy each dependency on its own).
+
+## Use Cases
+
+The following subsections will describe two different use cases how dependencies could be used.
+
+### (1) Need at least one "Provider"
+
+A package might want to declare that at least one dependency from a group needs to be present (called `satisfy-at-least-one`).
+Compared to "normal" dependencies it is not required that all dependencies of that group need to be present to satisfy the dependency group.
+
+An example for this is the `rmw_implementation` package.
+Currently it depends on all RMW implementation packages (e.g. `rmw_fastrtps_cpp`, `rmw_connext_cpp`, etc.) in order to ensure that it may only be built after all RMW implementation packages have been built.
+The goal would be that at least one RMW implementation needs to be present in order to build or use the `rmw_implementation` package.
+It would be acceptable if more than one or even all RMW implementations are available but that is not required.
+
+In case multiple package declare to be part of that dependency group the `rmw_implementation` package might want to provide an ordered list of preferred names which should be considered in that order.
+
+### (2) All packages of a specific "type"
+
+Another use case is that a dependency group doesn't need to be explicitly defined by all dependency names (called `satisfy-all`).
+Instead a dependency group is defined by an identifier and the expectation is that all packages which declare that they are being part of that group should be available.
+
+An example for this is the `ros1_bridge` package.
+Currently it depends on a manually selected set of message packages.
+The goal would be that it can declare that it depends on *all* packages which provide messages.
+
+Since the `ros1_bridge` manifest only contains the name of the group dependency, it doesn't control which packages declare that they are part of that group.
+It might want to exclude specific packages to be considered even though they are in the group.
+
+## Processes
+
+The following two processes need to be update to consider dependency groups.
```

Comment:
The buildfarm will also require changes won't it? It uses the rosdistro cache to determine build job ordering separately from the build or release tooling. 

Setting aside potential inconsistencies between package group members at bloom-time and group members in the current rosdistro cache:
- For `satisfy-all` we would need to trigger a build whenever any upstream group member is rebuilt.

- For `satisfy-at-least-one`, in the cases like rmw implementations, we would want all group members to be available at build time (bloom would likely list them all as Build Depends so that support for them is compiled in. Regardless of which is chosen at install-time) which means we can use the same behavior as `satisfy-all`.

But I don't see how the buildfarm would be able to choose which group members to include or not if there was a reason that all members could not be part of the same build (for example if two dependencies were mutually exclusive).

Reply:
I added mentioning `ros_buildfarm` in 0534486af1398a3410159130183ed4c918dcdbbd.

> But I don't see how the buildfarm would be able to choose which group members to include or not if there was a reason that all members could not be part of the same build (for example if two dependencies were mutually exclusive).

If some of the group members conflict with each other `satisfy-at-least-one` could fall back to remove a package from the set (instead of using `all`). Which one to be removed can be decided based on a preferred list and a deterministic order (e.g. alphabetical).

Reply:
@nuclearsandwich Out of curiousity, what situations are you thinking of with two mutually exclusive dependencies?  It's not that I can't imagine it ever happening, but more like I would mostly consider it a bug in the packaging/metadata if it did happen.  I'm just wondering in what scenarios you see it happening.

Reply:
typo: need to be update`d`